tests/__fixtures__/specfile_retries.spec.js

```diff
@@ -0,0 +1,33 @@
+console.log('Retry file loaded.')
+
+describe('retry', function () {
+    this.retries(2)
+
+    console.log('Retry suite loaded.')
+
+    it('should execute once due to zero retries', function () {
+        this.retries(0)
+
+        console.log('Test for zero retries loaded.')
+
+        throw Error('Deliberate error.')
+    })
+
+    it('should execute twice due to one test-level retry', function () {
+        this.retries(1)
+
+        console.log('Test for one retry loaded.')
+
+        throw Error('Deliberate error.')
+    })
+
+    it('should execute three times due to two suite-level retries', function () {
+        console.log('Test for two retries loaded.')
+
+        throw Error('Deliberate error.')
+    })
+
+    it('should run for at least 500ms to test reported duration', function () {
+        browser.pause(500)
+    })
+})
```

Comment:
This is testing mocha based retries. The smoke test should create a file and fail in the first run and check if that created file exist and pass in the second. Also we should move this to `/tests/mocha/retry.js`

Reply:
I'm not exactly sure how you want the tests restructured.

Here is a list of all the things that should be tested:
1. Test that mocha test-level and suite-level retries work as expected. This is technically not necessary since the folks at mocha will take care of this, however, it does make sense to just have a sanity check for the interaction of test-level retries and suite-level retries with specfile-level retries.
2. Test that specfile-level retries cause specs to be rerun until they succeed or run out of retries.
3. Test that retried specs that ultimately fail also cause the testrun as a whole to fail, and that retried specs that ultimately succeed do not cause the testrun as a whole to fail.
4. Test that reporters receive the correct `retries` property.
5. Test that reporters receive the correct duration of retried and succeeded runs.
6. Test that retries run with the correct runner id.
This is what I can think of now, feel free to add more.

To test this, we need to run a (mocha) test that exhibits all the possible behaviours. Call this Test A. It must be run at least twice, once to retry and fail, and once to retry and succeed. 

Then, the outcomes of the runs of Test A must be analyzed to see if the correct actions were taken and if the correct results were reported. Call this Test B.

This is the reason for having the two separate test files. Furthermore, to record the `_duration` and `retries` properties from Test A, a reporter must be attached which allows us to record those properties. This is the reason for the custom reporter. Furthermore, it seems like one cannot override the `reporters` argument for `launch` in the smoke test with a class reference as it gets lost during serialisation, which is the reason for the separate config file.

Do you agree with the concept of two separate tests? If not a custom reporter, how would one record the `_duration` and `retries` properties of runs for Test A? And if a custom reporter, it seems like the separate config file is necessary so the reporter class can be defined in the config file itself.

This is a mouthful... I hope I clarified properly. If I'm completely off, please clarify, otherwise, please suggest on the way forward.

Reply:
thanks for your explanation. Here are my thoughts:

- I don't think we need to test 1. at all - as you said this is handled by the mocha devs and we should trust them
- I agree we you that we should have a smoke test were the test passes after retry and where it still fails after retrying, I suggest to create two test files in `/tests/mocha`, one that has a simple test and always fails and another where it checks if a e.g. "retryMe" file exist in the same dir, if not than create that file and fail if yes than pass, then we create two launcher tests in `/tests/smoke.runner.js` where we run 2 tests using these files (re 2. and 3.)
- We should not check if reporters receive the retry or duration property, that should be captured on the unit test level (re 4. and 5.)
- I think it would be cool to check with the retry/passing smoke test if 2 wdio  logs were created (this will also indirect check if the runner id was passed in correctly)

To give some clarification of how the smoke tests are structured: the idea behind them is to run wdio testrunner tests that check specific things e2e (no unit tests / jest tests in there). It really should test things on a functional level (no details). All spec files for these testruner tests should be located in `/tests/mocha`, the testruns should be programmatically triggered in `/tests/smoke.runner.js`. The main config is located in `/tests/helpers/config.js` so that the smoke runner can always use that and overwrites specific properties based on the test scenario (e.g. spec files or reporters).

If you have further questions feel free to PM me on Gitter.